File: sdk/typescript/tests/transport/netplayTransport.test.ts
import { describe, expect, test } from "bun:test";
import {
  NetplayRestClient,
  NetplayWebSocketEndpoint,
  createRoomRequest,
  netplayPaths,
  type HttpMethod,
  type NetplayAuthHeadersProvider,
  type NetplayHttpRequest,
  type NetplayHttpResponse,
  type NetplayHttpTransport,
} from "../../src/index.ts";
import { roomView, sessionDescriptor } from "../support/fixtures.ts";

describe("TypeScript netplay transport helpers", () => {
  test("REST client signs and sends create room request", async () => {
    const transport = new CapturingTransport(
      JSON.stringify({
        room: roomView(),
      }),
    );
    const auth = new CapturingAuthHeadersProvider();
    const client = new NetplayRestClient(transport, auth);

    await client.createRoom(createRoomRequest(sessionDescriptor()));

    expect(transport.lastRequest?.method).toBe("POST");
    expect(transport.lastRequest?.pathAndQuery).toBe("/v1/rooms");
    expect(transport.lastRequest?.headers.Authorization).toBe("signed");
    expect(transport.lastRequest?.body).toContain('"desktopProtocolVersion":3');
    expect(auth.lastPath).toBe("/v1/rooms");
    expect(auth.lastBody).toBe(transport.lastRequest?.body);
  });

  test("WebSocket reconnect path includes escaped token and epoch", async () => {
    const auth = new CapturingAuthHeadersProvider();
    const endpoint = new NetplayWebSocketEndpoint(auth);

    const request = await endpoint.joinRequest({
      inviteCode: "AB CD",
      reconnectTicket: {
        playerIndex: 0,
        resumeToken: "resume token/+",
        roomEpoch: 4,
      },
      role: "host",
    });

    expect(request.pathAndQuery).toBe(
      "/v1/ws?inviteCode=AB%20CD&protocolVersion=3" +
        "&playerIndex=0&roomEpoch=4&resumeToken=resume%20token%2F%2B",
    );
    expect(auth.lastPath).toBe(request.pathAndQuery);
  });

  test("room status path escapes invite code", () => {
    expect(netplayPaths.roomStatus("AB CD")).toBe("/v1/rooms/AB%20CD/status");
  });
});

class CapturingTransport implements NetplayHttpTransport {
  public lastRequest: NetplayHttpRequest | null = null;

  public constructor(private readonly responseBody: string) {}

  public async execute(request: NetplayHttpRequest): Promise<NetplayHttpResponse> {
    this.lastRequest = request;

    return {
      body: this.responseBody,
      statusCode: 200,
    };
  }
}

class CapturingAuthHeadersProvider implements NetplayAuthHeadersProvider {
  public lastBody: string | null = null;
  public lastPath: string | null = null;

  public async headersFor(
    method: HttpMethod,
    pathAndQuery: string,
    body: string | null,
  ): Promise<Readonly<Record<string, string>>> {
    void method;
    this.lastBody = body;
    this.lastPath = pathAndQuery;

    return {
      Authorization: "signed",
    };
  }
}

File: sdk/kotlin/src/test/kotlin/app/shadowboy/netplay/sdk/transport/NetplayTransportTest.kt
package app.shadowboy.netplay.sdk.transport

import app.shadowboy.netplay.sdk.NetplayPaths
import app.shadowboy.netplay.sdk.json.NetplayJson
import app.shadowboy.netplay.sdk.protocol.CreateRoomRequest
import app.shadowboy.netplay.sdk.protocol.CreateRoomResponse
import app.shadowboy.netplay.sdk.protocol.roomJson
import app.shadowboy.netplay.sdk.protocol.testSessionDescriptor
import app.shadowboy.netplay.sdk.state.ReconnectTicket
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.encodeToString

class NetplayTransportTest {
    @Test
    fun `rest client signs and sends create room request`() = runTest {
        val transport = CapturingTransport(
            NetplayJson.format.encodeToString(
                CreateRoomResponse.serializer(),
                CreateRoomResponse(
                    room = NetplayJson.format.decodeFromString(
                        app.shadowboy.netplay.sdk.protocol.RoomView.serializer(),
                        roomJson(),
                    ),
                ),
            ),
        )
        val auth = CapturingAuthHeadersProvider()
        val client = NetplayRestClient(transport, auth)

        client.createRoom(CreateRoomRequest(session = testSessionDescriptor()))

        val request = assertNotNull(transport.lastRequest)
        assertEquals(HttpMethod.Post, request.method)
        assertEquals("/v1/rooms", request.pathAndQuery)
        assertEquals("signed", request.headers["Authorization"])
        assertTrue(requireNotNull(request.body).contains("\"desktopProtocolVersion\":4"))
        assertEquals(request.pathAndQuery, auth.lastPath)
        assertEquals(request.body, auth.lastBody)
    }

    @Test
    fun `websocket reconnect path includes escaped token and epoch`() = runTest {
        val auth = CapturingAuthHeadersProvider()
        val endpoint = NetplayWebSocketEndpoint(auth)

        val request = endpoint.joinRequest(
            inviteCode = "AB CD",
            role = WebSocketRole.Host,
            reconnectTicket = ReconnectTicket(
                playerIndex = 0,
                roomEpoch = 4,
                resumeToken = "resume token/+",
            ),
        )

        assertEquals(
            "/v1/ws?inviteCode=AB%20CD&protocolVersion=4&supportsStateFileRelay=false&supportsRomFileRelay=false&playerIndex=0&roomEpoch=4&resumeToken=resume%20token%2F%2B",
            request.pathAndQuery,
        )
        assertEquals(request.pathAndQuery, auth.lastPath)
    }

    @Test
    fun `input websocket path includes token and session epoch`() = runTest {
        val auth = CapturingAuthHeadersProvider()
        val endpoint = NetplayWebSocketEndpoint(auth)

        val request = endpoint.inputJoinRequest(
            NetplayInputWebSocketJoinOptions(
                inviteCode = "AB CD",
                playerIndex = 1,
                roomEpoch = 4,
                sessionEpoch = 7,
                inputSocketToken = "input token/+",
            ),
        )

        assertEquals(
            "/v1/ws/input?inviteCode=AB%20CD&protocolVersion=4&playerIndex=1&roomEpoch=4&sessionEpoch=7&inputSocketToken=input%20token%2F%2B",
            request.pathAndQuery,
        )
        assertEquals(request.pathAndQuery, auth.lastPath)
    }

    @Test
    fun `room status path escapes invite code`() {
        assertEquals("/v1/rooms/AB%20CD/status", NetplayPaths.roomStatus("AB CD"))
    }
}

private class CapturingTransport(private val responseBody: String) : NetplayHttpTransport {
    var lastRequest: NetplayHttpRequest? = null

    override suspend fun execute(request: NetplayHttpRequest): NetplayHttpResponse {
        lastRequest = request
        return NetplayHttpResponse(statusCode = 200, body = responseBody)
    }
}

private class CapturingAuthHeadersProvider : NetplayAuthHeadersProvider {
    var lastPath: String? = null
    var lastBody: String? = null

    override suspend fun headersFor(
        method: HttpMethod,
        pathAndQuery: String,
        body: String?,
    ): Map<String, String> {
        lastPath = pathAndQuery
        lastBody = body
        return mapOf("Authorization" to "signed")
    }
}
